state_transfer/quiesce.go

```diff
@@ -0,0 +1,639 @@
+package state_transfer
+
+import (
+	"context"
+	"encoding/json"
+	"strconv"
+	"time"
+
+	ocappsv1 "github.com/openshift/api/apps/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	batchv1beta "k8s.io/api/batch/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	NodeSelectorAnnotation = "migration.openshift.io/preQuiesceNodeSelector"
+	QuiesceNodeSelector    = "migration.openshift.io/quiesceDaemonSet"
+	ReplicasAnnotation     = "migration.openshift.io/preQuiesceReplicas"
+	SuspendAnnotation      = "migration.openshift.io/preQuiesceSuspend"
+)
+
+// Quiesce applications on source cluster
+func QuiesceApplications(cfg *rest.Config, ns string) error {
```

Comment:
I wired up this example to use the scale subresource instead of patching the objects on each GVR differently, if it helps you in anyway to make this modular https://gist.github.com/alaypatel07/4c1494f70b3cebad1f9534c1c0fde639#file-scale-go-L151. 

I would have strongly urged to do this way but I think it will not work with jobs and deploymentconfig is doubtful so it is kind of up to you.

Reply:
Update 1: I verified, deploymentconfigs works with generic approach in the gist

Reply:
@alaypatel07 thank you for investigating. I was talking to Shawn late yesterday and I think it would be nice if we had an interface for quiesce as well. We could have simple, scaler, and / or others. It may be beneficial to have multiple options like with the other interfaces; here I could even see potential for perhaps running multiple quiesce options on one cluster if we become able to handle different classes of applications better or worse.